Feature/camera

yuri-ash opened the pull request:
Added Camera node.

How to:
Scene has a ptr directly on current camera.
When scene is visited, if camera ptr is present, then we push a custom command to the renderer to update the projection matrix.
After the scene is done parsing all its children, a custom command is created to pop the projection matrix.
CameraTest has been added in cppTest project. You can move and zoom the camera with your mouse .

Known issue:
- Renderer::checkVisibility doesn't work properly anymore, as the perspective view is directly modified.
- Zoom will zoom over the whole render, as only  one camera is active at a time. Even if the UI can stay still by adding it to the scene graph as a camera children, there is no such trick for a node to not be affected by zoom. - FIXED
- Input are not yet transform in camera view, so picking doesnt work properly. - FIXED (more patched than fixed, but it's working)

Improvement:
- add a node ViewCamera that will handle camera operation, allowing to have multiple camera targetting specific branch of the scene graph (ie: to one ViewCamera, we add a game scene, to an other view camera, we ad dour UI, both ViewCamera have different camera and both are children of scene. While moving a camera, the other render should stay unchanged) - DONE
- I haven't code a camera in a while, so computing of projection matrix might be a bit mixed with the view matrix.

yuri-ash committed:
Add camera to cocos2d.
Add test to show camera usage.

yuri-ash committed:
Merge branch 'v3' into feature/camera

Conflicts:
	cocos/2d/cocos2d.vcxproj
	tests/cpp-tests/proj.win32/cpp-tests.vcxproj.filters

yuri-ash committed:
add camera view to handle camera.
remove camera from scene.

minggo commented:
@yuri-ash 
We plan to add Camera in v3.3.
@ricardoquesada, does this PR fit our needs?

yuri-ash committed:
use scale for zoom in camera.
remove useless parameter in getCamera.

yuri-ash committed:
add a recursive function to get node / parent node as cameraView.
Add transform function callback in touch, to allow different camera.
Set transform function during event dispatcher, based on current node.
Not the best way to transform input into camera view, need a rework.

yuri-ash commented:
@minggo 
I saw that you were planning on adding it but i needed it right away for my project. I just submit this pull request if it can help you go a bit faster as I don't know your status on this feature.

Commit:
add camera source to makelist.
fix crash in event handler when camera view has no camera.

mindryu commented:
@yuri-ash 
I've find this. 
thank you.

Commit:
add input in camera.
fix crash in event handler because of camview/camera.
add camera in android build.
